### PR TITLE
[url_launcher_linux] file resource URI support for canLaunch

### PR DESCRIPTION
**Pull request by jokerttu**
Fixes `canLaunch` implementation for `file` scheme. 
`canLaunch` incorrectly returned a `false` response to launchable file scheme URLs.
Fix uses same method than [g_app_info_launch_default_for_uri](https://gitlab.gnome.org/GNOME/glib/-/blob/main/gio/gappinfo.c#L791-811) does internally to detect default application for URI.

Fixes [#93326](https://github.com/flutter/flutter/issues/93326)

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Note that unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I [updated pubspec.yaml](https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates) with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[the auto-formatter]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code

### CODE REVIEW

**Comment by jokerttu** [comment]
Seems like new tests are failing on `Cirrus CI`.
It is likely that the Cirrus CI system will not be able to open that filepath used on test (`file:/`) and returns always false.
Unit tests will be rewritten after fix for [#88724](https://github.com/flutter/flutter/issues/88724) is available in `stable`.

Mockable system apis are needed to be able to mock file check; and I even used a bit of my time to test one approach to implement those ([commit](https://github.com/CodemateLtd/plugins/commit/3ce4a7a0c09d35ae22ee8a7c907c12f7668c1c4d)), but desided to wait fix for work already done related to [#88724](https://github.com/flutter/flutter/issues/88724)

**Comment by stuartmorgan** [comment]
Thanks for the submission! Marking as a draft per the comment above; please remove the draft designation once the tests have been reworked.

File: packages/url_launcher/url_launcher_linux/CHANGELOG.md
Comment (stuartmorgan) [inline packages/url_launcher/url_launcher_linux/CHANGELOG.md:3]: Nit: Fixes

(Per the repo's new CHANGELOG style guide)

File: packages/url_launcher/url_launcher_linux/linux/url_launcher_plugin.cc
Comment (stuartmorgan) [inline packages/url_launcher/url_launcher_linux/linux/url_launcher_plugin.cc:54]: Why isn't this an autoptr?

File: packages/url_launcher/url_launcher_linux/linux/url_launcher_plugin.cc
Comment (stuartmorgan) [inline packages/url_launcher/url_launcher_linux/linux/url_launcher_plugin.cc:58]: Same.

File: packages/url_launcher/url_launcher_linux/linux/url_launcher_plugin.cc
Comment (stuartmorgan) [inline packages/url_launcher/url_launcher_linux/linux/url_launcher_plugin.cc:80]: I would expect that a function called `can_launch_url_with_file_scheme` would expect its argument to be a URL with a file scheme, but that's not being checked anywhere AFAICT.

File: packages/url_launcher/url_launcher_linux/linux/test/url_launcher_linux_test.cc
Comment (stuartmorgan) [inline packages/url_launcher/url_launcher_linux/linux/test/url_launcher_linux_test.cc:44]: Shouldn't this be `file:///`? I'm surprised this works.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

**Comment by stuartmorgan** [comment]
Adding @robert-ancell for second review.

File: packages/url_launcher/url_launcher_linux/linux/test/url_launcher_linux_test.cc
Comment (robert-ancell) [inline packages/url_launcher/url_launcher_linux/linux/test/url_launcher_linux_test.cc:45]: This is a minor memory leak, which the existing tests are doing. I've made https://github.com/flutter/plugins/pull/4648 to fix this, so it probably should also be fixed in this PR.

**Review by robert-ancell — CHANGES_REQUESTED** [review]
Should be good with the requested changes, thanks @jokerttu.

File: packages/url_launcher/url_launcher_linux/linux/test/url_launcher_linux_test.cc
Comment (jokerttu) [reply packages/url_launcher/url_launcher_linux/linux/test/url_launcher_linux_test.cc:44]: This is proper mimimal representation of the file scheme URI pointin to local linux root folder:
> The minimal representation of a local file with no authority field and an absolute path that begins with a slash "/"

[rfc8089#appendix-B](https://datatracker.ietf.org/doc/html/rfc8089#appendix-B)

File: packages/url_launcher/url_launcher_linux/CHANGELOG.md
Comment (jokerttu) [reply packages/url_launcher/url_launcher_linux/CHANGELOG.md:3]: There was colliding commit having version ## NEXT by @stuartmorgan
I kept it as it was, but added this change to as new line.

File: packages/url_launcher/url_launcher_linux/linux/url_launcher_plugin.cc
Comment (jokerttu) [reply packages/url_launcher/url_launcher_linux/linux/url_launcher_plugin.cc:80]: This is actually mimicking how `g_app_info_launch_default_for_uri` detects if URI is launchable. 
Internally glibc checks first if there is registered schema for URI and if not, it checks if URI contains file that can be launched. [Source for g_app_info_launch_default_for_uri](https://github.com/GNOME/glib/blob/main/gio/gappinfo.c#L809-L821)


glibc can support [multiple file schemes](https://github.com/GNOME/glib/blob/main/gio/gvfs.c#L208-L214) additional to `file`-scheme, for example [`resource:`](https://github.com/GNOME/glib/blob/main/gio/gvfs.c#L118-L120)-scheme, and these should be checked as well. 
So it might be that in some use cases URI won't be starting with `file`-scheme, but is launchable anyway.
For this reason I think it is important that canLaunch always return true for URIs that are  launchable with launch method.

File: packages/url_launcher/url_launcher_linux/linux/test/url_launcher_linux_test.cc
Comment (jokerttu) [reply packages/url_launcher/url_launcher_linux/linux/test/url_launcher_linux_test.cc]: I did this fix only for this `file:` scheme test, cos other tests are fixed within #4648.

File: packages/url_launcher/url_launcher_linux/linux/url_launcher_plugin.cc
Comment (jokerttu) [reply packages/url_launcher/url_launcher_linux/linux/url_launcher_plugin.cc]: Function renamed to `can_launch_uri_with_file_resource`

File: packages/url_launcher/url_launcher_linux/CHANGELOG.md
Comment (stuartmorgan) [inline packages/url_launcher/url_launcher_linux/CHANGELOG.md:1]: This needs to match the new version (which is why `publish` is failing).

File: packages/url_launcher/url_launcher_linux/linux/url_launcher_plugin.cc
Comment (stuartmorgan) [inline packages/url_launcher/url_launcher_linux/linux/url_launcher_plugin.cc:59]: These 5 lines collapses to `return app_info != nullptr;` now that you are using autoptrs

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

File: packages/url_launcher/url_launcher_linux/linux/url_launcher_plugin.cc
Comment (stuartmorgan) [inline packages/url_launcher/url_launcher_linux/linux/url_launcher_plugin.cc:48]: Nit: s/Check/Checks/, and add the missing period, to follow Google style.

**Review by stuartmorgan — APPROVED** [review]
LGTM with a comment nit.

This will need final approval from @robert-ancell as well.

**Review by robert-ancell — APPROVED** [review]
Thanks @jokerttu!

File: packages/url_launcher/url_launcher_linux/linux/url_launcher_plugin.cc
Comment (jokerttu) [reply packages/url_launcher/url_launcher_linux/linux/url_launcher_plugin.cc]: Fixed

**Pull request merged**
